setup_machine.sh

```diff
@@ -18,17 +18,21 @@ set -e
 
 sudo yum update -y
 
-# # https://tecadmin.net/install-python-3-9-on-centos/
+# https://tecadmin.net/install-python-3-9-on-centos/
 sudo yum install gcc openssl-devel bzip2-devel libffi-devel zlib-devel -y
 sudo yum install wget -y
-wget https://www.python.org/ftp/python/3.9.7/Python-3.9.7.tgz
-tar xzf Python-3.9.7.tgz
-cd Python-3.9.7
-sudo ./configure --enable-optimizations
-sudo make altinstall
-cd ..
-sudo rm Python-3.9.7.tgz*
 
+# Check if Python aleady installed. If already there, skips next 7 lines.
+if [[ $(python3.9 --version) != "Python 3.9.7" ]]; then
```

Comment:
You should quote the process substitution, i.e.

```
if [[ "$(python3.9 --version)" != "Python 3.9.7" ]]; then
```

The quotes guard against the possibility that `$(python3.9 --version)` produces nothing. Without the quotes, an empty result will cause a syntax error in your script, which would be confusing (`shellcheck` would have caught this I think)

Reply:
Ah, I just learned something I didn't know. You are safe because you used the double-square-brackets `[[ ... ]]`.

I put this little test script in a file:

```
#!/bin/bash

if [[ $(python --version) != "Python 3.9.7" ]]; then
    echo nope
fi
```

and `shellcheck` doesn't complain about it. I ran it and there is no syntax error and it prints `nope` -- my Python is version 2.7.18 so the body of the `if` runs.

But when I change the `[[ ... ]]` to single brackets `[ ... ]`  like this:

```
#!/bin/bash

if [ $(python --version) != "Python 3.9.7" ]; then
    echo nope
fi
```

then `shellcheck` does complain:

```
MacBook-Air:tmp kcharter$ shellcheck junk.sh

In junk.sh line 3:
if [ $(python --version) != "Python 3.9.7" ]; then
     ^-----------------^ SC2046 (warning): Quote this to prevent word splitting.

For more information:
  https://www.shellcheck.net/wiki/SC2046 -- Quote this to prevent word splitt...
```

and if I try running the script it fails with a syntax error:

```
MacBook-Air:tmp kcharter$ ./junk.sh
Python 2.7.18
./junk.sh: line 3: [: !=: unary operator expected
```

Reply:
To be on the safe side, you should redirect standard error to standard output when you run `python3 --version` like so:

```
$(python3 --version 2>&1)
```

The reason is that some versions of Python will print the version on standard error rather than standard output. The Python 2 on my Mac does this. The Python 3 I have (3.9.9) prints it on standard output so your script will probably work, but I think it's a good idea to include the redirection just to be safe.

I learned this by trying a little experiment with redirecting the standard output to `/dev/null`. For `python`, which is Python 2.7.18 on my machine, I still see the version when I redirect standard output to `/dev/null`:

```
MacBook-Air:tmp kcharter$ python --version > /dev/null
Python 2.7.18
MacBook-Air:tmp kcharter$
```

but for `python3` I don't:

```
MacBook-Air:tmp kcharter$ python3 --version > /dev/null
MacBook-Air:tmp kcharter$
```

This shows that `python --version` is printing to standard error, but `python3 --version` is printing to standard output. Maybe they changed this in Python 3 and all Python 3 versions will print the version on standard output, but it's hard to say for sure.

You can make sure you capture the version string no matter what by redirecting standard error to standard output with `2>&1`.

Reply:
a. Every time I ran shellcheck, it never flagged needing double quotes. I'm pretty sure this is why almost everything I read advises to use `[[ ]]` around a comparison. I've added " " around `"$(python3.9 --version)"` mostly since the line looks more syntactically correct/balanced in the comparison:
`if [[ "$(python3.9 --version)" != "Python 3.9.7" ]]; then'

b. Since I state in the comparison to check for the version, python3.9, I'm going to leave out your suggestion of adding in 2>&1. From most of what I read, this is only an issue for Python versions Python 3.5 or earlier & since I was specific to a much later version, I won't have this issue.